feat: Users are no longer deleted immediately, but marked as deleted.…

jekutzsche opened the pull request:
… The marked users can no longer be used and are no longer displayed. However, the data is still available, for example, for working with the audit logs. After all references to the users have been deleted in due time, the users are also permanently deleted.

- Entity and database extended by the fields necessary for soft-delete and user locking.
- The repository now contains methods that take the `deletedAt` flag into account.
- Extends the `loadUserByUsername` method to also take into account the `enabled` and `locked` flags.
- Adds the `locked` status to the handling of an update. This allows the status to be set initially with the insert and changed later via the update.
- A job has been added to permanently delete users as soon as this is possible. If there is an error with the referential integrity, then this is ignored for a configurable period of time, since data such as the audit logs must be stored for a certain time and have references.
- User mapper extended to also map from `UserInsertDTO` to `UserAccount` and manual mapping removed.
- Code of the DTOs rewritten to Records and thus greatly simplified.
- `InitialAdminLoader` has been moved to the `users` package, because this class has more to do with user management than with authorization.

Refs iris-connect/iris-backlog#235

jekutzsche committed:
feat: Users are no longer deleted immediately, but marked as deleted. The marked users can no longer be used and are no longer displayed. However, the data is still available, for example, for working with the audit logs. After all references to the users have been deleted in due time, the users are also permanently deleted.

- Entity and database extended by the fields necessary for soft-delete and user locking.
- The repository now contains methods that take the `deletedAt` flag into account.
- Extends the `loadUserByUsername` method to also take into account the `enabled` and `locked` flags.
- Adds the `locked` status to the handling of an update. This allows the status to be set initially with the insert and changed later via the update.
- A job has been added to permanently delete users as soon as this is possible. If there is an error with the referential integrity, then this is ignored for a configurable period of time, since data such as the audit logs must be stored for a certain time and have references.
- User mapper extended to also map from `UserInsertDTO` to `UserAccount` and manual mapping removed.
- Code of the DTOs rewritten to Records and thus greatly simplified.
- `InitialAdminLoader` has been moved to the `users` package, because this class has more to do with user management than with authorization.

Refs iris-connect/iris-backlog#235

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/users/entities/UserAccount.java:
```diff
@@ -44,6 +45,21 @@ public class UserAccount extends Aggregate<UserAccount, UserAccountIdentifier> {
+	public UserAccount markDeleted() {
+
+		setDeletedAt(Instant.now());
+		setUserName(UUID.randomUUID().toString());
```
This is a good idea.

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/users/UserAccountsRepository.java:
```diff
@@ -4,12 +4,21 @@
+interface UserAccountsRepository extends JpaRepository<UserAccount, UserAccountIdentifier> {
 
-	long countByRole(UserRole role);
+	Optional<UserAccount> findByUserNameAndDeletedAtIsNull(String userName);
```
You are absolutely right in principle! Even if we have to start with `@Query`, the readability of the method calls is more important, thanks.

I would rename all methods according to a scheme. However, I am not yet happy with `findActiveUser...`, since users can also be locked, so they are not "active", but should be returned by the methods.

jekutzsche replied:
As discussed, I omit "Active", resulting in a name like `findUserByUsername`.

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/users/UserDeleteJob.java:
```diff
@@ -0,0 +1,68 @@
+	@Scheduled(cron = "${iris.client.user.delete-cron:-}")
+	void deleteUsers() {
+
+		var deletedUsers = users.findAllByDeletedAtIsNotNull();
```
This is fine, thanks.

jekutzsche committed:
chore: improves the code according to the comments in the PR

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/users/entities/UserAccount.java:
```diff
@@ -44,6 +45,21 @@ public class UserAccount extends Aggregate<UserAccount, UserAccountIdentifier> {
+	public UserAccount markDeleted() {
+
+		setDeletedAt(Instant.now());
+		setUserName(getId().toString());
```
What's the reason for replacing the username with the id when soft deleting a user?

jekutzsche replied:
The username is unique and will be released for new use by replacing it. This is mainly to prevent confusion when "deleted" users would block usernames.

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/users/UserDeleteJob.java:
```diff
@@ -0,0 +1,66 @@
+
+			try {
+
+				users.delete(it);
```
Deleting a user as soon as possible and relying only on the audit log date could be dangerous.
- What if there are / going to be different storage time requirements for users and audit logs (e.g. by law)?
- What if there are side effects (bugs or code changes) in the audit log functionality that triggers a user deletion to soon or to late (or not at all)?

Does ist make sense to add an additional isAfter check in combination with a dedicated userDeleteAfter property. The default value of this property could be equal to the getDeleteErrorWarningAfter prop.

jekutzsche replied:
We rely on references to a user with the current implementation. Thus, the user remains as long as it is referenced, even with future new records with different storage time.

I don't see a retention period for the user record itself, since storage must always have a justification and I don't see that in the user data alone.

To late (or not at all) user deletion should stand out due to the logging. To soon could become a problem in case of a programming error. Do we make such? ;)

> Does ist make sense to add an additional isAfter check in combination with a dedicated userDeleteAfter property. The default value of this property could be equal to the getDeleteErrorWarningAfter prop.

This raises whether it is legal to keep the user data if it is no longer referenced anywhere. But for simplicity's sake, maybe we'd better discuss that in the Daily.

jekutzsche replied:
As documentation:

As it turned out in the Daily, the assumption that users can be deleted as soon as there is no more reference was wrong. E.g. `IrisMessageFolder` and `UserAccounts` themselves also have a reference to users and since they have no deletion period, they can prevent deletion. Deletion is thus inconvenient and has been replaced by anonymization. The check if references exist has been limited to certain relevant references. Furthermore, a configurable time-based anonymization has been implemented.

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/users/UserDeleteJob.java:
```diff
@@ -0,0 +1,66 @@
+			} catch (DataIntegrityViolationException e) {
+
+				if (now().minus(properties.getDeleteErrorWarningAfter()).isAfter(it.getDeletedAt())) {
+					log.warn("User with ID = {} could not be deleted!", it.getId());
```
What happens / has to happen in this case?

jekutzsche replied:
Nothing is currently planned except logging. If this occurs, the deadlines are too long or we have done something wrong. 

This one is a compromise between avoiding expectable error logs and preventing errors from remaining invisible.

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/users/UserAccountsRepository.java:
```diff
@@ -4,12 +4,29 @@
-	long countByRole(UserRole role);
+	String SELECT_BASE = "SELECT u from UserAccount u where ";
+
+	@Query(SELECT_BASE + "u.userName = :userName AND u.deletedAt IS NULL")
```
Basically, it is double that the users are no longer found, but it is intended that way. The "deleted" users are no longer directly usable. The records should only be used to be able to resolve the data to the references (IDs).

At least that was the idea behind it, which may well be debatable.

jekutzsche committed:
chore: re-implements delete job as anonymization job and integrates reference checkers

As it turned out in the Daily, the assumption that users can be deleted as soon as there is no more reference was wrong. E.g. `IrisMessageFolder` and `UserAccounts` themselves also have a reference to users and since they have no deletion period, they can prevent deletion. Deletion is thus inconvenient and has been replaced by anonymization. The check if references exist has been limited to certain relevant references. Furthermore, a configurable time-based anonymization has been implemented.

jekutzsche committed:
chore: improves access level of the `UserAnonymizationReferenceChecker` implementations

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/cases/CaseDataRequestRepository.java:
```diff
@@ -34,4 +35,7 @@ public interface CaseDataRequestRepository extends JpaRepository<CaseDataRequest
 
 	Optional<CaseDataRequest> findByIdOrDataAuthorizationToken(DataRequestIdentifier id, String dataAuthorizationToken);
+
+	@Query("select count(r)>0 from CaseDataRequest r where r.metadata.createdBy = :userId OR r.metadata.lastModifiedBy = :userId")
```
We could avoid rewriting the same code for every single user-referenced entity by creating and extending a dedicated "AggregateRepository" (which itself extends the JpaRepository). The AggregateRepository would come with the method "isReferencedToUser".

Furthermore we could move every meta-data related method to the "AggregateRepository". E.g.: findByMetadataCreatedIsBefore

jekutzsche replied:
I think I was able to take up the suggestion sensibly and thus simplify the code. Thanks for the food for thought.

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/users/web/dto/UserRoleDTO.java:
```diff
@@ -4,5 +4,5 @@
  */
 public enum UserRoleDTO {
-	ADMIN, USER
+	ADMIN, USER, DELETED
```
Do we need the DELETED role in the Dto? Deleted users should never show op in the frontend, right?

jekutzsche replied:
In fact, according to the current application, we do not need this. This was just the simplest solution to satisfy MapStruct when creating the mappers.

jekutzsche replied:
Alternatively, I need to tell MapStruct to which value the `DELETED` from the entity should be mapped.

jekutzsche committed:
chore: introduces `AggregateRepository` as the base repository for aggregate-level operations and removes duplicate methods in each repository

jekutzsche commented:
:tada: This PR is included in version 1.6.0-rc.1 :tada:

The release is available on [GitHub release](https://github.com/iris-connect/iris-client/releases/tag/v1.6.0-rc.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

jekutzsche commented:
:tada: This PR is included in version 1.6.0 :tada:

The release is available on [GitHub release](https://github.com/iris-connect/iris-client/releases/tag/v1.6.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
